src/main/scala/chisel3/util/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,234 @@
+  fileName match {
+    case SuffixRegex(_*) =>
+      Builder.warning(
+        s"""LoadMemoryAnnotation fileName "$fileName" has extension ".txt" will still be appended"""
+      )
+    case _ =>
+  }
+
+  def transformClass : Class[LoadMemoryTransform] = classOf[LoadMemoryTransform]
+
+  def toFirrtl: LoadMemoryAnnotation = LoadMemoryAnnotation(target.toNamed, fileName, hexOrBinary)
+}
+
+/**
+  * Firrtl implementation for load memory
+  * @param target        memory to load
+  * @param fileName      name of input file
+  * @param hexOrBinary   use $readmemh or $readmemb
+  */
+case class LoadMemoryAnnotation(
+  target: Named,
+  fileName: String,
+  hexOrBinary: String = "h",
+  originalMemoryNameOpt: Option[String] = None
+) extends SingleTargetAnnotation[Named] {
+
+  def getFileName: String = {
+    (target, originalMemoryNameOpt) match {
+      case (ComponentName(componentName, _), Some(originalMemoryName)) =>
+        if(componentName == originalMemoryName) {
+          fileName
+        }
+        else {
+          fileName + componentName.drop(originalMemoryName.length)
+        }
+      case _ =>
+        fileName
+    }
+  }
+
+  def duplicate(n: Named): LoadMemoryAnnotation = {
+//    if(n.serialize != target.serialize) {
+////      Builder.errors.error(s"LoadMemory[${target.serialize}] must be a simple memory, no bundles")
+//      throw new Exception(s"LoadMemory[${target.serialize}] must be a simple memory, no bundles")
+//    }
+    target match {
+      case ComponentName(componentName, _) =>
+        this.copy(target = n, originalMemoryNameOpt = Some(componentName))
+      case _ =>
+        this.copy(target = n)
+    }
+  }
+}
+
+/**
+  * This pass creates BlackBoxInlineAnno from the LoadMemoryAnnotations
+  * it does this even if the backend is not verilog.
+  *
+  * @param circuitState the target circuit state
+  */
+//TODO: (chick) support a treadle or interpreter means of memory loading
+//TODO: (chick) can this only be done when backend is known to support this.
+//scalastyle:off method.length cyclomatic.complexity regex
+class CreateBindableMemoryLoaders(circuitState: CircuitState) extends Pass {
+
+  val annotations      : Seq[Annotation] = circuitState.annotations
+  val memoryAnnotations: Seq[LoadMemoryAnnotation] = annotations.collect{ case m: LoadMemoryAnnotation => m }
+
+  val bindModules      : mutable.ArrayBuffer[BlackBoxInlineAnno] = new mutable.ArrayBuffer()
+
+  val verilogEmitter   : VerilogEmitter = new VerilogEmitter
+
+  /**
+    * finds the specified module name in the circuit
+    *
+    * @param moduleName name to find
+    * @param circuit circuit being analyzed
+    * @return the circuit, exception occurs in not found
+    */
+  def findModule(moduleName: String, circuit: Circuit): DefModule = {
+    circuit.modules.find(module => module.name == moduleName) match {
+      case Some(module: firrtl.ir.Module) =>
+        module
+      case Some(externalModule: DefModule) =>
+        externalModule
```

Comment:
I am not actually sure what's going on here. I don't see why a DefModule would correspond to external.  This some clear cargo cult programming and despite having done it myself I don't know where I got it from.  I'll try and puzzle this out.

Reply:
Yeah for the record, this could be written as:
```scala
circuit.modules.find(module => module.name == moduleName)
       .getOrElse(throw new Exception(s"Could not find module $moduleName in circuit $circuit"))
```
Or it could just match on `DefModule`. alternatively, if there are going to be lots of lookups (or really more than like 2), you could instead just create a map and query that:
```scala
val moduleMap: Map[String, DefModule] = circuit.modules.map(m => m.name -> m).toMap
```

Reply:
@chick Is there any resolution to this? I don't see any reason to have this function since you can just do `circuit.modules.find(module => module.name == moduleName).get` anywhere this would be used. In any case, the matching on`Module` and `DefModule` is unnecessary

Reply:
For the record, I'm not sure if Github is bugging out here or what but I had to click `Load more` on `hidden conversations` to see this. That makes it look kind of like the `outdated` conversations`. I don't think this is outdated though, it's just not obvious from the PR paged.

Reply:
@jackkoenig I am removing the findModules, it is not used